### ISSUE TITLE
Relative imports

### ISSUE THREAD

**Issue by ccbogel** [ccbogel/QualCoder#500]
@maralorn  IMPORTANT
I cannot run the code from command line or via PyCharm using relative imports.

This is one reason why I  removed all the relative imports from a previous developer. I will do so again unless there is a good solution. I can use if else blocks around the imports (e.g. try importing relative then try importing absolute), but I do not think this is a great solution.

**Comment by maralorn** [comment]
How are you trying to run the code?

**Comment by maralorn** [comment]
I am sure that PyCharm has to have a way to work together with setuptools.

For running from command line, this is the way to go (only sure about this on linux systems):

Run `python setup.py install`. After that you can run `qualcoder` when the directory `setup.py` installs to is in your PATH variable. Otherwise run `path/to/qualcoder`.

If you want to immediately try out every change you can instead use `python setup.py develop`. (https://setuptools.readthedocs.io/en/latest/userguide/development_mode.html).

That way the `qualcoder` program is not a copy of your code but just a link into your working directory, so everytime you reexecute the program it uses the current state of your code.

**Comment by ccbogel** [comment]
OK. I have python knowledge in some areas, but not in others.
I used this successfully to install via develop:, (and then qualcoder from the command line on Ubuntu 20.04):
sudo python3 setup.py develop

**Comment by ccbogel** [comment]
Also. Does this mean the instillation instructions will have to change in the Readme? as this setup.py is now working better (thank you).
So for windows:
py setup.py install
Linux:
python3 setup.py install
Mac:
python setup.py install

**Comment by ccbogel** [comment]
In PyCharm, I have marked qualcoder folder as the source root
![Screenshot from 2021-06-29 20-08-50](https://user-images.githubusercontent.com/5834638/123779896-fa312f00-d915-11eb-9ab6-d3ad202b45a7.png)

**Comment by mweinelt** [comment]
Right, try to bulk replace `from .` with `from qualcoder.` to make these imports absolute. It is likely the better choice anyway.

**Comment by ccbogel** [comment]
Danke, Ich wuerde probieren

**Comment by maralorn** [comment]
My problem is, that I am also not a python expert and I feel a bit bad about sending you down a rabbit hole in which you could get lost.

But let’s try to tackle this one by one:




> I used this successfully to install via develop:, (and then qualcoder from the command line on Ubuntu 20.04):
> sudo python3 setup.py develop

Great!



> I find relative imports unclear. Do absolute imports not work with setup.py ?

I am actually not sure, why exactly I used `import .<foo>` everywhere. I thought that might be better in certain circumstances where the package is not configured correctly as library.
But I suspect now that you should use `import qualcoder.<foo>` everywhere and that could probably fix some of your issues (together with running `python3 setup.py develop`.)
That is the way how normal python projects usually do this, so I am pretty sure it’s the way to go. Sorry, for leading you astray there. If you want I can make a PR to fix that.

> I have googled but do not see a solution to use in PyCharm.

Yeah, not sure either. I’ve never used PyCharm. But I have people I can ask, if the problem persists. I only found https://www.jetbrains.com/help/pycharm/creating-and-running-setup-py.html, but maybe if we do the above mentioned fix, this problem will solve itself?

> Also. Does this mean the instillation instructions will have to change in the Readme?

Yes, probably.

> So for windows:
> py setup.py install
> Linux:
> python3 setup.py install
> Mac:
> python setup.py install

The Linux command is correct. For the other two I really don‘t know. It might be that they just keep working, so I would first try that. The next thing I would try is to change the conditional so that you use the `entry_points` option on mac and windows, too. Then you can generate the exe by using `python3 setup.py py2exe` on windows e.g.

> And the Debian I create may not work with the relative imports?

Yeah, I think using `python3 setup.py install` in the debian install might work better.

> And double-clicking on the qualcoder.py or the shortcut to it on Windows, may not work?

Likely. Maybe it will work after running `setup.py install`.

**Comment by ccbogel** [comment]
@mweinelt latest error after updating to qualcoder.foo

![pic1](https://user-images.githubusercontent.com/5834638/123786902-93177880-d91d-11eb-8544-0da0dfe3574d.png)

**Comment by mweinelt** [comment]
In the Run/Debug configuration, make sure you configure the correct module.

**Comment by ccbogel** [comment]
I dont know how to change additional configurations, apart from what is shown here:
![pic1](https://user-images.githubusercontent.com/5834638/123790511-c3611600-d921-11eb-97fa-01ba3b1098ed.png)

Also. QualCoder does not work on Window now. I have a work computer, with little battery left. The command line setup does not work and double-click on the qualcoder.py does not work. So I will have to add try: except: to have the option for absolute imports

**Comment by ccbogel** [comment]
Currently the relative imports are not working on Windows.
Add in a try, except to allow the option of importing relative or absolute now creates a new error which I cannot resolve:
ImportError: cannot import name 'DialogReportCodeSummary' from partially initialized module 'report_code_summary' (most likely due to a circular import) (/home/cc/Documents/informatics/QualCoder-master/qualcoder/report_code_summary.py)

Solution for tomorrow. I will remove the relative imports.

**Comment by maralorn** [comment]
If you can‘t get it to work now and you have little time left I suggest you checkout an older working commit or do a `git revert` on the commits that are problematic for you. They are still documented in the history and here on github so you can apply them again when we have found solutions for all ensuing problems.

I didn‘t want to cause any trouble, sorry. My expectation was that you would only merge my pull request if it works for you.

Still it might be nice to fix this in the future because it would simplify installing the software for users and make it user to package for other linux distributions.

**Comment by fmagin** [comment]
Quick guess why the PyCharm behavior is confusing: You have "Add Source Roots to PYTHONPATH" activated in the PyCharm Run configuration (which is the default). This adds the source roots (blue folders in the PyCharm project tree) to the path Python uses to find modules. This will not be the case when running from the command line. The cleanest and easily reproducible way for a development setup is to install the package with `pip install -e ./folder_with_setup_py/`, then there will be a `qualcoder` package that is just a link to the development directory, i.e. changes in PyCharm will also be used on the next python run, without having to reinstall.

**Comment by fmagin** [comment]
Also, make sure to open the folder that contains the `setup.py` in PyCharm, i.e. the root of the git repository, this might be causing some issues though I am not sure if it is the issue in this case. But it at least makes the git integration work nicely.

**Comment by ccbogel** [comment]
@fmagin Thank you. On Ubuntu PyCharm works and comman line run works.
On Windows I get the following errors:
<img width="663" alt="pic1" src="https://user-images.githubusercontent.com/5834638/123882157-8a5c8c00-d989-11eb-854f-3252bf26d55e.png">

**Comment by fmagin** [comment]
I do not expect `py -m __main__.py` to ever work, `-m` expects a _module_ not a _file_, hence why `py -m qualcoder` works.

The failure of `py __main__.py` could just be that `qualcoder` isn't installed in that python environment, did you run `pip install ./QualCoder` ( the folder with the `setup.py`) before? If you want changes to the source to be reflected without reinstalling you can also use `pip install --editable ./QualCoder` (or `-e` for short).

I am not familiar with `py2exe`, did this produce a proper executable before?
My first attempt would be to change the lines
https://github.com/ccbogel/QualCoder/blob/2fc1500ef0db1bde277840d03f55f616fafcb5c2/setup.py#L25-L29
to 
```python
elif sys.platform == 'win32':
     extra_options = dict(
         setup_requires=['py2exe'],
         options = {'py2exe': {'bundle_files': 1, 'compressed': True}},
         windows = [{'script': mainscript}],
     )
```

and see if that works. I don't have a proper windows machine to test this on, so try this locally and check if it works.

**Comment by ccbogel** [comment]
I tried that. It still only produces an egg in the dist folder.
However, double-clicking the __main__.py file ran the software.

**Comment by fmagin** [comment]
yeah, double clicking `__main__.py` probably just runs `python __main__.py`, and if the python environment has the package installed via `pip` it should work. That would still require users to run `pip install QualCoder` before though.

Did you try `setup.py py2exe` ? `setup.py install` just installs it to the current python environment, that isn't really supposed to create a standalone `.exe` anyway

**Comment by ccbogel** [comment]
I did not do pip install Qualcoder on Windows - it just worked. This was the same before, when there were all absolute imports, double click worked.
I have tried the above commands, and an exe was not produced.
Anyway - that is for future work. The main thing is that the relative imports are working on Ubuntu and Windows at least.
Later I will try pyinstaller again, I did have some brief success with pyinstaller  before.
